api/envoy/extensions/filters/http/waf/v3/waf.proto

```diff
@@ -0,0 +1,223 @@
+//          action:
+//            reset: truek
+//    - name: ofac-white-list
+//      match:
+//        sourceIp:
+//          cidrs:
+//          - addressPrefix: 2.144.136.0
+//            prefixLength: 21
+//          - addressPrefix: 2.144.144.0
+//            prefixLength: 20
+//          - addressPrefix: 2.144.160.0
+//            prefixLength: 19
+//      watcher:
+//        - name: envoy.watcher
+//          action:
+//            nop:true
+//    - name: ofac-black-list
+//      match:
+//        andRules:
+//          rules:
+//          - header:
+//              name: ":path"
+//              prefix: "/ofac-page"
+//              invertMatch: true
+//          - sourceIp:
+//              cidrs:
+//              - addressPrefix: 2.144.136.0
+//                prefixLength: 21
+//              - addressPrefix: 2.144.144.0
+//                prefixLength: 20
+//              - addressPrefix: 2.144.160.0
+//                prefixLength: 19
+//      watcher:
+//        - name: envoy.watcher
+//          action:
+//            http_redirect:
+//              query_params: >
+//                "name1=value1&ref=%REQ(X-REQUEST-ID)%&name3=val3"
+//              response_code: "302"'
+//              host: "www.acme.com"'
+//      rule_match_cb: envoy.until_first_match
+//  - name: grp_2
+//    policies:
+//	  - name: ciruit-breaker
+//    	match:
+//      	header: {name: "path", prefix_match: "/signup"}
+//    	watcher:
+//      	name: custom.circuit_breaker
+//      	config:
+//        	type : gradient2
+//        	action: drop
+//	 - name: bot-analyzer
+//     match:
+//       header: {name: "path", prefix_match: "/"}
+//     watcher:
+//       name: custom.bot_analyzer
+//   rule_match_cb: envoy.all
+
+message WAF {
+  message PolicySet {
+    repeated Policy policies = 1[(validate.rules).repeated .min_items = 1];
+  }
+
+  message PolicyGroup {
+    // Name of the Policy Group.
+    string name = 1 [(validate.rules).string.pattern = "^[a-zA-Z0-9]([-a-zA-Z0-9]*[a-zA-Z0-9])?$"];
+    PolicySet policies = 2;
+    RuleMatchCB callback = 3;
+  }
+
+  repeated PolicyGroup policy_groups = 1[(validate.rules).repeated .min_items = 1];
+  Executor executor = 2;
+ }
+
+ // Rule defines a match condition and an associated action.
+ message Policy {
+  string name = 1 (validate.rules).string.pattern = "^[a-zA-Z0-9]([-a-zA-Z0-9]*[a-zA-Z0-9])?$"];
+  Match match = 2;
+  Watcher watcher = 3;
+ }
+
+ // Executor that will execute the actions based on priority or another
+ // custom logic
+
+
+ message Executor {
+   string name = 1;
+   google.protobuf.Any config = 2;
+ }
+
+ // RuleMatchCB that will decide if the rule engine should
+ // 1. Continue evaluating remaining rules that are bound to this callback
+ // 2. Possibly apply request header mutations that Watchers may depend on
+ // 3. Pass along the actions recommended by watchers to others
+
+ message RuleMatchCB {
+   string name = 1;
+   google.protobuf.Any config = 2;
```

Comment:
I'm not sure the notion of "callback" makes a ton of sense at the API level. Do you mean to say `MatchExtension`?